### PR TITLE
Fix layout broadcasting arithm ops

### PR DESCRIPTION
**Pull request by stiepan**
## Category:
**Bug fix** (*non-breaking change which fixes an issue*)

## Description:
Without this change, it is possible to produce tensors whose layout does not match their dimensionality.

The arithm ops layout propagation works by picking just any non-empty layout and making sure that any other operands have either empty or equal layout. This leads to two issues:
* layout of wrong length: if more ndim-ed tensor has no layout and the less ndim-ed tensor does, the output ends up with layout of incorrect lengt. Take `a` of shape (1, 2, 3, 4) and empty layout, and  `b` of shape  (4,) and layout "C". The output ends up to have 4dims and layout "C".
* errors when the layout match but are not equal, for example FHWC and C.

The fix is to 1. check if the picked layout is not too short and if so, discard it, 2. allow for different layouts as long as one if suffix of the other.

## Additional information:
It would be nice to follow-up the layout propagation with more involved mechanism that would:
1. not discard partial information, i.e. when bigger dim-ed tensor does not have layout, but the samller dim-ed has, we should retain the information about layout suffix. Possibly with "*" wildcards in the layout (that would have consitently special treatment in equality). 
2. generalize the matching from common suffix to one of the layouts being subsequence of the other.
### Tests:

- [ ] Existing tests apply
- [x] New tests added
  - [x] Python tests
  - [ ] GTests
  - [ ] Benchmark
  - [ ] Other
- [ ] N/A

### Documentation
- [x] Existing documentation applies
- [ ] Documentation updated
  - [ ] Docstring
  - [ ] Doxygen
  - [ ] RST
  - [ ] Jupyter
  - [ ] Other
- [ ] N/A

#### Requirements
- [ ] Implements new requirements
- [ ] Affects existing requirements
- [x] N/A

**REQ IDs**: N/A

**JIRA TASK**: DALI-3556

### CODE REVIEW

File: dali/test/python/operator_1/test_arithmetic_ops.py [hunk 1]
```diff
@@ -863,6 +864,118 @@
...
+        expected_layout = expected_layout or ""
+        assert o.layout() == expected_layout, f"got `{o.layout()}`, expected `{expected_layout}`"
+        out_shape = o.shape()
+        assert len(out_shape) == batch_size, f"got `{len(out_shape)}`, expected `{batch_size}`"
+        for sample_shape in out_shape:
+            assert sample_shape == expected_shape, \
+                f"got `{sample_shape}`, expected `{expected_shape}`"
+    else:
+        with assert_raises(Exception, glob="They must be equal or one must be a suffix"):
+            o, = p.run()
```
[the remark this answers is not in the document]
Comment (stiepan) [reply dali/test/python/operator_1/test_arithmetic_ops.py]: done

**Comment by stiepan** [comment]
!build

**Comment by stiepan** [comment]
!build

File: dali/operators/math/expressions/arithmetic.h [hunk 2]
```diff
@@ -84,6 +84,20 @@ inline TileCover GetOneTilePerSample(const TensorListShape<> &shape) {
   return std::make_tuple(std::move(descs), std::move(ranges));
 }
 
+/**
+ * @brief Checks if the child is a (possibly improper) suffix of the parent.
+ * The child must not be longer than the parent.
+ */
+inline bool HasSuffix(const TensorLayout &parent, const TensorLayout &child) {
```
Comment (mzient) [inline dali/operators/math/expressions/arithmetic.h:91]: More common name: EndsWith
```suggestion
inline bool EndsWith(const TensorLayout &parent, const TensorLayout &child) {
```

File: dali/operators/math/expressions/arithmetic.h [hunk 3]
```diff
@@ -84,6 +84,20 @@ inline TileCover GetOneTilePerSample(const TensorListShape<> &shape) {
...
+  assert(parent.size() >= child.size());
```
Comment (mzient) [inline dali/operators/math/expressions/arithmetic.h:92]: Unnecessary and actually harmful. Obviously a longer string is not a suffix of a shorter one.
```suggestion
  if (parent.size() < child.size())
    return false;
```

File: dali/operators/math/expressions/arithmetic.h (hunk 2 above)
Comment (stiepan) [reply dali/operators/math/expressions/arithmetic.h]: done

File: dali/operators/math/expressions/arithmetic.h (hunk 3 above)
Comment (stiepan) [reply dali/operators/math/expressions/arithmetic.h]: done

**Comment by stiepan** [comment]
!build

File: dali/operators/math/expressions/arithmetic.h [hunk 4]
```diff
@@ -330,9 +346,22 @@ class ArithmeticGenericOp : public Operator<Backend> {
...
     auto curr_batch_size = ws.GetInputBatchSize(0);
 
+    PropagateShapes<Backend>(result_shape_, *expr_, ws, curr_batch_size);
+
     if (!types_layout_inferred_) {
       result_type_id_ = PropagateTypes<Backend>(*expr_, ws);
       result_layout_ = GetCommonLayout<Backend>(*expr_, ws);
+      // sample_ndim is assumed to be constant between iterations.
+      if (result_layout_.size() != result_shape_.sample_dim()) {
+        // If the largest dimentional tensors do not have layout set but some
```
Comment (mzient) [inline dali/operators/math/expressions/arithmetic.h:356]:
```suggestion
        // If the tensors with the most dimensions do not have a layout set but some
```

File: dali/operators/math/expressions/arithmetic_test.cc [hunk 5]
```diff
@@ -134,7 +134,7 @@ TEST(ArithmeticOpsTest, TreePropagationLayoutError) {
     ptr->Resize({{1}, {2}}, DALI_INT32);
   }
   in[0]->SetLayout(TensorLayout());
-  in[1]->SetLayout(TensorLayout("HW"));
+  in[1]->SetLayout(TensorLayout("DH"));
```
Comment (mzient) [inline dali/operators/math/expressions/arithmetic_test.cc:137]: Actually, we should test both cases - the old one should _not_ throw, but instead extend the layout to `"DHW"`.

Comment (stiepan) [reply dali/operators/math/expressions/arithmetic_test.cc:137]: We have tests for that in python though.

File: dali/test/python/operator_1/test_arithmetic_ops.py [hunk 6]
```diff
@@ -863,6 +864,118 @@ def get_sh(arg_idx):
...
+        placement.append("gpu" if placement[-1] == "cpu" else "cpu")
+        return tuple(placement)
+
+    def get_input_types(num_inputs, integral_only):
+        types = (np.int32, np.uint8)
+        if not integral_only:
+            types += (np.float32, )
+        return tuple(rng.choice(types, size=(num_inputs, )))
+
+    bin_layouts = [
```
Comment (mzient) [inline dali/test/python/operator_1/test_arithmetic_ops.py:884]: An explanation would be nice.
```suggestion
    # The input layouts and the expected output layout.
    # A number N denotes an ND tensor without a layout.
    # `Exception` means that applying an operator with the arguments with given layouts should raise an error.
    bin_layouts = [
```

File: dali/test/python/operator_1/test_arithmetic_ops.py [hunk 7]
```diff
@@ -863,6 +864,118 @@ def get_sh(arg_idx):
...
+        (2, bin_layouts, bin_ops),
+        (3, ternary_layouts, ternary_operations),
+    ]:
+        for i, (args_desc, out_desc) in enumerate(layouts):
+            assert (len(args_desc) == num_inputs)
+            op, op_name = op_lists[i % len(op_lists)][:2]
+            op = op if not isinstance(op, tuple) else op[0]
+            input_devs = get_input_dev(num_inputs)
+            in_types = get_input_types(num_inputs, op_name in ("&|^"))
+            args_desc = tuple(tensor_desc(arg) for arg in args_desc)
```
Comment (mzient) [inline dali/test/python/operator_1/test_arithmetic_ops.py:942]: :+1: for not creating a useless Cartesian product

File: dali/operators/math/expressions/arithmetic.h [hunk 8]
```diff
@@ -84,6 +84,21 @@ inline TileCover GetOneTilePerSample(const TensorListShape<> &shape) {
   return std::make_tuple(std::move(descs), std::move(ranges));
 }
 
+/**
+ * @brief Checks if the child is a (possibly improper) suffix of the parent.
+ */
+inline bool EndsWith(const TensorLayout &parent, const TensorLayout &child) {
```
[the remark this answers is not in the document]
Comment (stiepan) [reply dali/operators/math/expressions/arithmetic.h]: done

File: dali/test/python/operator_1/test_arithmetic_ops.py (hunk 6 above)
Comment (stiepan) [reply dali/test/python/operator_1/test_arithmetic_ops.py:888]: done

File: dali/operators/math/expressions/arithmetic.h (hunk 4 above)
Comment (stiepan) [reply dali/operators/math/expressions/arithmetic.h]: done

**Comment by stiepan** [comment]
!build

**Review by mzient — APPROVED** [review]

**Review by szalpal — APPROVED** [review]

**Pull request merged**